### ISSUE TITLE
There should be a warning for repeated variables even in curried arguments

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6070]
**Original bug ID:** 6070
**Reporter:** kaustuv
**Status:** acknowledged (set by @damiendoligez on 2013-07-17T14:57:26Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 4.00.1
**Category:** typing
**Monitored by:** @gasche @hcarty

## Bug description

I've just spent half an hour bug hunting an issue that came from code of this form (greatly simplified):

    type error = E of string
    let first_msg x (E x) = x

Obviously, this function is equivalent to fun _ (E x) -> x, which is silently the wrong interpretation of my intentions.

Basically, I would have spotted the error right away if instead I had written

    let first_msg (x, E x) = x

but, since first_msg was being used in fold_left, I had to write it curried. The nature of this bug in my code took me on a complete wild goose chase as I hadn't assumed that the bug would be in my printing routine for errors.

Can there be a warning for shadowed variables in the arguments of curried function definitions? Note that this needs to be different from warning 27, which is way too annoying even to turn on, far less mark as error.

**Comment by vicuna** [comment]
**Comment author:** @alainfrisch

Currently, "fun p1 p2 -> e" is treated exactly as "fun p1 -> fun p2 -> e".  What would be the exact criterion to trigger the warning?  Should it collect such immediately nested abstractions and check that they bind disjoint set of variables?  What about the following:

 let f x ?(a = x) x = a + x

Here we have a duplicated variable, but the instances are both used.

Personally, I find warning 27 quite useful (and it is turned into an error in LexiFi's code base).  What's wrong with it?

**Comment by vicuna** [comment]
**Comment author:** kaustuv

I think that if the user writes "fun x x -> x", then it is nearly always something strange and possibly unintentonal. This is also true for "fun x ?(a=x) x -> x".

Warning 27 also complains about such as "let snd x y = y" and if it is an error then I have to constantly fiddle with initial underscore. I am sure one can get used to it, but I would rather differentiate between "repeated uses of variable in function arguments" and "unused variable in function arguments" as potential sources of errors in my code.

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

Maybe the criterion could be "the variable is shadowed before it is ever used", without regard for the shape of the expression. I don't much like a special warning just for function definitions.

**Comment by vicuna** [comment]
**Comment author:** kaustuv

I think I am OK with that interpretation.

In fact, that could just be a new case of "suspicious unused variable" (warning 26). Although, enlarging the scope of existing warnings risks breaking existing code that mark those warnings as errors.

**Comment by garrigue** [comment]
How could we implement @damiendoligez 's suggestion?
It would be natural to have warning 26 (enabled by default) in that case (shadowing of an unused variable). I.e., this is not innocuous.

**Comment by shindere** [comment]
What shall we do according to you, @garrigue?

My understanding of the issue is that there is something we would like
to tackle, so the issue needs to stay open, but we don't know yet how to
best tackle it. Do you (and others) share this understanding?

One other question: is the original author still interested, nine years
later?

**Comment by garrigue** [comment]
I have no strong opinion on this one. There is already a warning for unused variables, which would apply in this case, so this could be a sufficient answer. @damiendoligez would be nicer for a warning enabled by default, but this is extra work for something that is already covered by another warning.

**Comment by shindere** [comment]
Okay I am going to close this and it can always be re-opened if somebody
feels the need to.
